### PR TITLE
SNMP Module Support

### PR DESCRIPTION
**Pull request by meenakshidembi691**
# Description
SNMP Module Support

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, pep8, linting, or security issues
- [x] I have performed Ansible Sanity test using --docker default
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Unit Test
- [x] Functional Test

### CODE REVIEW

File: plugins/modules/snmp_manager.py [hunk 1]
```diff
@@ -330,10 +331,71 @@ def delete_snmp_manager(self, snmp_manager_id):
             LOG.error(msg)
             self.module.fail_json(msg=msg, **utils.failure_codes(e))
 
+    def validate_version_change(self, snmp_manager_params, snmp_manager_details):
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:334]: Does the API not handle this?

File: playbooks/modules/snmp_manager.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        version: "V3"
+        alert_severity: Critical
+        trap_community: test
+        snmp_username: test
+        auth_protocol: MD5
+        auth_privacy: TDES
+        auth_pass: Password123!
+        state: present
+
+    - name: Get SNMP Manager
```
Comment (trisha-dell) [inline playbooks/modules/snmp_manager.yml:42]: Are we supporting get?

File: playbooks/modules/snmp_manager.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- name: SNMP Manager operations on Powerstore Storage System
+  hosts: localhost
+  connection: local
+  vars:
+    array_ip: "10.**.**.**"
+    validate_certs: false
+    user: "admin"
+    password: "*****"
+
+  tasks:
```
Comment (trisha-dell) [inline playbooks/modules/snmp_manager.yml:11]: Please add a modification task as well.

File: plugins/modules/snmp_manager.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+
+DOCUMENTATION = r'''
+---
+
+module: snmp_manager
+version_added: '3.6.0'
+short_description: Manage SNMP Managers for PowerStore
+description:
+- Managing SNMP Managers on PowerStore Storage System includes creating
+  a SNMP Manager, getting details of SNMP Manager, modifying a
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:17]: Getting details should ideally be supported only from Info

File: plugins/modules/snmp_manager.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+  new_ip_address:
+    description:
+    - IP address or FQDN of the SNMP server to update.
+    - IPv4 and IPv6 are supported.
+    type: str
+  snmp_port:
+    description:
+    - Port number to use with the address of the SNMP server.
+    type: int
+  snmp_password:
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:65]: this should come before snmp_port as it needs to be alphabetical.

File: plugins/modules/snmp_manager.py [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+    type: str
+  state:
+    description:
+    - Define whether the file DNS should be enabled or not.
+    - For Delete operation only, it should be set to C(absent).
+    choices: ['absent', 'present']
+    type: str
+    default: 'present'
+
+notes:
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:103]: Aren't we handling this using attribute format?

File: plugins/modules/snmp_manager.py [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+    description: Whether or not the resource has changed.
+    returned: always
+    type: bool
+    sample: "false"
+
+snmp_details:
+    description: Details of the SNMP manager.
+    returned: When SNMP exists.
+    type: complex
+    contains:
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:183]: arrange this alphabetically.

File: plugins/modules/snmp_manager.py [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+
+        if auth_protocol != "None" and version == 'V3':
+            payload['authpass'] = create_params.get('snmp_password')
+
+        return payload
+
+    def create_snmp_manager(self, create_params):
+        """Create SNMP Server"""
+        try:
+            msg = 'Attempting to create SNMP Server'
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:298]: server -> manager, applicable for all instances.

File: plugins/modules/snmp_manager.py [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+        auth_protocol=dict(type='str', choices=['Nil', 'MD5', 'SHA256']),
+        ip_address=dict(type='str', aliases=['network_name'], required=True),
+        new_ip_address=dict(type='str'),
+        snmp_port=dict(type='int'),
+        snmp_password=dict(type='str', aliases=['auth_pass'], no_log=True),
+        snmp_username=dict(type='str'),
+        trap_community=dict(type='str'),
+        update_password=dict(default='always', type='str', choices=['always', 'on_create']),
+        version=dict(type='str', choices=['V2c', 'V3']),
+        state=dict(default='present', type='str', choices=['present', 'absent'])
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:439]: type=str is default, it an be removed.

File: plugins/modules/snmp_manager.py [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+        if snmp_manager_params['state'] == 'present' and not snmp_manager_details:
+            snmp_manager_details = snmp_obj.create_snmp_manager(create_params=snmp_manager_params)
+            snmp_obj.result['changed'] = True
+
+        SNMPManagerModifyHandler().handle(snmp_obj, snmp_manager_params, snmp_manager_details)
+
+
+class SNMPManagerHandler():
+    def handle(self, snmp_obj, snmp_manager_params):
+        snmp_manager_details = None
```
Comment (Bhavneet-Sharma) [inline plugins/modules/snmp_manager.py:480]:
```suggestion
        snmp_manager_details = dict()
```

File: plugins/modules/snmp_manager.py [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+
+def get_powerstore_snmp_manager_parameters():
+    """This method provides the parameters required for the ansible
+    SNMP modules on PowerStore"""
+    return dict(
+        alert_severity=dict(type='str', choices=['Info', 'Minor', 'Major', 'Critical']),
+        auth_privacy=dict(type='str', choices=['Nil', 'AES256', 'TDES']),
+        auth_protocol=dict(type='str', choices=['Nil', 'MD5', 'SHA256']),
+        ip_address=dict(type='str', aliases=['network_name'], required=True),
+        new_ip_address=dict(type='str'),
```
Comment (Bhavneet-Sharma) [inline plugins/modules/snmp_manager.py:432]: Add an alias for this param

File: plugins/modules/snmp_manager.py (hunk 1 above)
Comment (meenakshidembi691) [reply plugins/modules/snmp_manager.py:359]: This is done to handle error messages better.

File: plugins/modules/snmp_manager.py [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+        payload = {
+            'port': create_params['snmp_port'],
+            'version': version,
+            'trap_community': trap_community if version != 'V3' else None,
+            'alert_severity': create_params['alert_severity'],
+            'ip_address': create_params['ip_address'],
+        }
+        if version == 'V3':
+            payload['user_name'] = create_params.get('snmp_username')
+            if auth_protocol == "Nil" or auth_protocol is None:
```
[the remark this answers is not in the document]
Comment (meenakshidembi691) [reply plugins/modules/snmp_manager.py:287]: This is done to handle error messages better.

File: plugins/modules/snmp_manager.py [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+author:
+- Meenakshi Dembi (@dembim) <ansible.team@dell.com>
+
+extends_documentation_fragment:
+  - dellemc.powerstore.powerstore
+
+options:
+  alert_severity:
+    description:
+    - Possible severities.
```
[the remark this answers is not in the document]
Comment (meenakshidembi691) [reply plugins/modules/snmp_manager.py:27]: This is as per the details of the API documentation.

File: plugins/modules/snmp_manager.py [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+    choices: ['Info', 'Minor', 'Major', 'Critical']
+    type: str
+  auth_privacy:
+    description:
+    - Supported SNMP privacy protocol.
+    - C(Nil) - No encryption on the wire.
```
[the remark this answers is not in the document]
Comment (meenakshidembi691) [reply plugins/modules/snmp_manager.py:34]: This is as per the details of the API documentation.

File: plugins/modules/snmp_manager.py [hunk 15]
```diff
@@ -0,0 +1,553 @@
...
+            errormsg = "Trap community is required parameter for creating SNMP Manager with version V2c."
+            LOG.error(errormsg)
+            self.module.fail_json(msg=errormsg)
+        if create_params['version'] == 'V3' and not (create_params['snmp_username'] and len(create_params['snmp_username'].strip()) > 0):
+            errormsg = "snmp_username is required parameter for creating SNMP Manager with version V3."
+            LOG.error(errormsg)
+            self.module.fail_json(msg=errormsg)
+        if (create_params['version'] == 'V3' and create_params['auth_protocol']) and \
+                not (create_params['snmp_password'] and len(create_params['snmp_password'].strip()) > 0):
+            errormsg = "snmp_password/auth_pass required parameter for creating or modifying SNMP Manager with version V3 and auth_protocol."
```
Comment (Bhavneet-Sharma) [inline plugins/modules/snmp_manager.py:465]:
```suggestion
            errormsg = "snmp_password or auth_pass required parameter for creating or modifying SNMP Manager with version V3 and auth_protocol."
```

File: plugins/modules/snmp_manager.py [hunk 16]
```diff
@@ -0,0 +1,553 @@
...
+
+    def common_validations(self, create_params):
+        if create_params['version'] == 'V3' and (create_params['auth_privacy'] and create_params['auth_privacy'] != "Nil") \
+                and (not create_params['auth_protocol'] or create_params['auth_protocol'] == "Nil"):
+            errormsg = "For V3 SNMP auth_protocol with None value must have privacy_protocol with None value."
+            LOG.error(errormsg)
+            self.module.fail_json(msg=errormsg)
+        if create_params['version'] == 'V3' and (not create_params['auth_protocol'] or create_params['auth_protocol'] == "Nil") \
+                and (not create_params['auth_privacy'] or create_params['auth_privacy'] == "Nil") and create_params['snmp_password']:
+            errormsg = "V3 with no authenticaton should not use snmp_password/authpass."
```
Comment (Bhavneet-Sharma) [inline plugins/modules/snmp_manager.py:477]:
```suggestion
            errormsg = "V3 with no authenticaton should not use snmp_password or authpass."
```

File: tests/unit/plugins/module_utils/mock_snmp_manager_api.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
+# Copyright: (c) 2022, Dell Technologies
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/module_utils/mock_snmp_manager_api.py:1]:
```suggestion
# Copyright: (c) 2024, Dell Technologies
```

File: tests/unit/plugins/module_utils/mock_snmp_manager_api.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+"""Mock Api response for Unit tests of DNS module for PowerStore"""
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/module_utils/mock_snmp_manager_api.py:5]:
```suggestion
"""Mock Api response for Unit tests of SNMP Manager module for PowerStore"""
```

File: tests/unit/plugins/modules/test_snmp_manager.py [hunk 19]
```diff
@@ -0,0 +1,285 @@
+# Copyright: (c) 2024, Dell Technologies
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+"""Unit Tests for storage container module for PowerStore"""
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/modules/test_snmp_manager.py:5]:
```suggestion
"""Unit Tests for SNMP manager module for PowerStore"""
```

File: tests/unit/plugins/modules/test_snmp_manager.py [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+
+import pytest
+# pylint: disable=unused-import
+from ansible_collections.dellemc.powerstore.tests.unit.plugins.module_utils.libraries import initial_mock
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/modules/test_snmp_manager.py:14]: If not used, remove the unused imports.

File: tests/unit/plugins/modules/test_snmp_manager.py [hunk 21]
```diff
@@ -0,0 +1,285 @@
...
+
+    @pytest.fixture
+    def module_object(self):
+        return PowerStoreSNMPManager
+
+    def test_get_snmp_manager(self, powerstore_module_mock):
+        self.set_module_params(
+            powerstore_module_mock, self.get_module_args,
+            {
+                "ip_address": "127.0.0.1"
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/modules/test_snmp_manager.py:38]: Check for Sonar and checkmarx issues.

File: playbooks/modules/snmp_manager.yml
Comment (sachin-apa) [inline playbooks/modules/snmp_manager.yml:1]: Lets add examples of how to create snmp for V3 with No Auth,  Auth Only, Auth with privacy

**Review by sachin-apa — APPROVED** [review]

File: tests/unit/plugins/modules/test_snmp_manager.py (hunk 21 above)
Comment (meenakshidembi691) [reply tests/unit/plugins/modules/test_snmp_manager.py:36]: this is dummy IP so sonar is not giving any issue with this

**Review by Bhavneet-Sharma — APPROVED** [review]
Update the RST file as well

File: plugins/modules/snmp_manager.py [hunk 22]
```diff
@@ -0,0 +1,553 @@
...
+    default: 'always'
+  version:
+    description:
+    - Supported SNMP protocol versions.
+    - C(V2c) - SNMP version 2c.
+    - C(V3) - SNMP version 3.
+    choices: ['V3', 'V2c']
+    type: str
+    default: 'V3'
+  state:
```
Comment (trisha-dell) [inline plugins/modules/snmp_manager.py:98]: This is not alphabetically arranged.

**Review by Bhavneet-Sharma — APPROVED** [review]

**Review by trisha-dell — APPROVED** [review]

**Review by sachin-apa — APPROVED** [review]

**Pull request merged**